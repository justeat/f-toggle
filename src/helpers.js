import $ from 'qwery';

/**
 * Handles the toggle element click events
 *
 * @param {string} toggleClass
 *
 * @property {object} toggle Toggle an element's visibility.
 * @property {object} show Shows an element.
 * @property {object} hide Hides an element.
 */
const toggles = toggleClass => ({
    toggle (element) {
        element.classList.toggle(toggleClass);
    },

    show (element) {
        element.classList.remove(toggleClass);
    },

    hide (element) {
        element.classList.add(toggleClass);
    }
});

/**
 * Handles the toggle element click events
 *
 * @param {string|string[]} targets
 * @param {string} toggleClass
 */
const handleToggles = (targets, toggleClass) => {
    if (!Array.isArray(targets)) {
        handleToggles(targets.split(' '), toggleClass);
        return;
    }

    targets.forEach(target => {
        const parts = target.split(':');

        if (parts.length === 1) {
            parts.unshift('toggle');
        }

        const [toggleType, toggleName] = parts;

        $(`[data-toggle-name~=${toggleName}]`)
            .forEach(toggles(toggleClass)[toggleType]);
    });

};

/**
 * Toggles the target you have clicked, and hides all other elements in the accordion
 *
 * @param {string} target
 * @param {object} accordion
 */

const handleAccordionToggles = (target, accordion) => {

    const toggleClass = accordion.getAttribute('data-toggle-class') || 'is-hidden';

    $('[data-toggle-name]', accordion)
        .filter(toggle => !toggle.hasAttribute('data-toggle-accordion-exclude'))
        .forEach(element => {
            const type = element.getAttribute('data-toggle-name') === target ? 'toggle' : 'hide';
            toggles(toggleClass)[type](element);
        });

};

export {
    toggles,
    handleToggles,
    handleAccordionToggles
}
