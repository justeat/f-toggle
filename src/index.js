/**
 * @overview Fozzie vanilla JS toggle library.
 *
 * @module toggle
 */

 import $ from 'qwery';


/**
 * Handles the toggle element click events
 *
 * @property {object} toggle Toggle an element's visibility.
 * @property {object} show Shows an element.
 * @property {object} hide Hides an element.
 */
const toggles = toggleClass => ({
    toggle (element) {
        element.classList.toggle(toggleClass);
    },

    show (element) {
        element.classList.remove(toggleClass);
    },

    hide (element) {
        element.classList.add(toggleClass);
    }
});

/**
 * Handles the toggle element click events
 *
 * @param {string|string[]} targets
 */
const handleToggles = (targets, toggleClass) => {
    if (!Array.isArray(targets)) {
        handleToggles(targets.split(' '), toggleClass);
        return;
    }

    targets.forEach(target => {
        const parts = target.split(':');

        if (parts.length === 1) {
            parts.unshift('toggle');
        }

        const [ toggleType, toggleName ] = parts;

        $(`[data-toggle-name~=${toggleName}]`)
            .forEach(toggles(toggleClass)[toggleType]);
    });

};

/**
 * Bind the toggle element click events
 */
export default () => {
    $('[data-toggle-target]')
        .forEach(toggle => {
            toggle.addEventListener('click', e => {
                e.preventDefault();

                const target = toggle.getAttribute('data-toggle-target');
                const toggleClass = toggle.getAttribute('data-toggle-class') || 'is-hidden';

                handleToggles(target, toggleClass);
            });
        });
};
