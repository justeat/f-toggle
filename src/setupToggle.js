import $ from 'qwery';
import closest from 'closest';
import { toggles, handleToggles, handleAccordionToggles } from './helpers';

export default () => {

    /**
     * If accordion, only display first section on initialisation
     */

    $('[data-toggle-accordion]')
        .forEach(accordion => {
            const toggleClass = accordion.getAttribute('data-toggle-class') || 'is-hidden';

            $('[data-toggle-name]', accordion)
                .filter(toggle => !toggle.hasAttribute('data-toggle-accordion-exclude'))
                .slice(1)
                .forEach(toggles(toggleClass).hide);

        });

    /**
     * Bind the toggle element click events
     */

    $('[data-toggle-target]')
        .forEach(toggle => {
            toggle.addEventListener('click', e => {
                e.preventDefault();

                const target = toggle.getAttribute('data-toggle-target');
                const toggleClass = toggle.getAttribute('data-toggle-class') || 'is-hidden';
                const accordionExclude = toggle.hasAttribute('data-toggle-accordion-exclude');
                const accordion = closest(toggle, '[data-toggle-accordion]');

                if (accordion && !accordionExclude) {
                    handleAccordionToggles(target, accordion);
                    return;
                }

                handleToggles(target, toggleClass);
            });
        });

};
