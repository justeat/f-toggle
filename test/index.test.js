import TestUtils from 'js-test-buddy';
import { setupToggle } from '../src/index';

describe('setupToggle', () => {

    it('is a function', () => {
        expect(typeof setupToggle).toBe('function');
    });

    describe('toggle method', () => {

        it('adds hidden class', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <button data-toggle-target="test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();
        });

        it('removes hidden class', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test" class="is-hidden"></div>
            <button data-toggle-target="test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('adds custom class when specified', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <button data-toggle-target="test" data-toggle-class="toggled"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('removes custom class when specified', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test" class="toggled"></div>
            <button data-toggle-target="test" data-toggle-class="toggled"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('can handle multiple targets', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <div data-toggle-name="test" class="is-hidden"></div>
            <button data-toggle-target="test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('can handle multiple specified targets', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="one" class="is-hidden"></div>
            <div data-toggle-name="two"></div>
            <button data-toggle-target="one two"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

    });

    describe('show method', () => {

        it('does nothing when element is already shown', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <button data-toggle-target="show:test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('removes hidden class', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test" class="is-hidden"></div>
            <button data-toggle-target="show:test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('removes custom class', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test" class="toggled"></div>
            <button data-toggle-target="show:test" data-toggle-class="toggled"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('can handle multiple targets', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <div data-toggle-name="test" class="is-hidden"></div>
            <button data-toggle-target="show:test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('can handle multiple specified targets', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="one" class="is-hidden"></div>
            <div data-toggle-name="two" class="is-hidden"></div>
            <button data-toggle-target="show:one show:two"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

    });

    describe('hide method', () => {

        it('does nothing when element is already hidden', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test" class="is-hidden"></div>
            <button data-toggle-target="hide:test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('adds hidden class', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <button data-toggle-target="hide:test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('adds custom class', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <button data-toggle-target="hide:test" data-toggle-class="toggled"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('can handle multiple targets', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test"></div>
            <div data-toggle-name="test" class="is-hidden"></div>
            <button data-toggle-target="hide:test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('can handle multiple specified targets', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="one"></div>
            <div data-toggle-name="two"></div>
            <button data-toggle-target="hide:one hide:two"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

    });

    describe('mixed toggles', () => {

        it('shows and hides elements correctly', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="one"></div>
            <div data-toggle-name="two" class="is-hidden"></div>
            <div data-toggle-name="three" class="is-hidden"></div>
            <div data-toggle-name="four"></div>
            <button data-toggle-target="one two show:three hide:four"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            TestUtils.click(button);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

    });

    describe('accordion', () => {

        it('should display only the first item on initialisation', () => {

            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-accordion data-toggle-first-section-expanded>
                <div data-toggle-name="one"></div>
                <button data-toggle-target="one"></button>
                <div data-toggle-name="two"></div>
                <button data-toggle-target="two"></button>
            </div>
        `);

            // Act
            setupToggle();

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('should have all items collapsed on initialisation', () => {

            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-accordion>
                <div data-toggle-name="one"></div>
                <button data-toggle-target="one"></button>
                <div data-toggle-name="two"></div>
                <button data-toggle-target="two"></button>
            </div>
        `);

            // Act
            setupToggle();

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('should update the button selected class on selecting new section', () => {

            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-accordion>
                <div data-toggle-name="one"></div>
                <button data-toggle-target="one"></button>
                <div data-toggle-name="two"></div>
                <button data-toggle-target="two"></button>
            </div>
        `);
            const [, button2] = document.querySelectorAll('button');

            // Act
            setupToggle();
            TestUtils.click(button2);

            // Assert
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('should hide all other open elements, and toggle the element clicked', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-accordion>
                <div data-toggle-name="one"></div>
                <button data-toggle-target="one"></button>
                <div data-toggle-name="two"></div>
                <button data-toggle-target="two"></button>
                <div data-toggle-name="three"></div>
                <button data-toggle-target="three"></button>
                <div data-toggle-name="four"></div>
                <button data-toggle-target="four"></button>
            </div>
        `);
            const [, button2, button3] = document.querySelectorAll('button');

            // Act & Assert
            setupToggle();
            TestUtils.click(button3);
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Act & Assert
            TestUtils.click(button2);
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('should not apply accordion behaviour to elements with \'data-toggle-accordion-exclude\'', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-accordion data-toggle-class="accordion--hide">
                <div data-toggle-name="one"></div>
                <button data-toggle-target="one"></button>
                <div data-toggle-name="two" data-toggle-accordion-exclude></div>
                <button data-toggle-target="two" data-toggle-accordion-exclude></button>
                <div data-toggle-name="three"></div>
                <button data-toggle-target="three"></button>
                <div data-toggle-name="four"></div>
                <button data-toggle-target="four"></button>
            </div>
        `);
            const [, button2, button3] = document.querySelectorAll('button');

            // Act & Assert
            setupToggle();
            TestUtils.click(button3);
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Act & Assert
            TestUtils.click(button2);
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

        it('should not apply accordion behaviour to elements with \'data-toggle-accordion-exclude\'', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-accordion data-toggle-class="accordion--hide">
                <div data-toggle-name="one"></div>
                <button data-toggle-target="one"></button>
                <div data-toggle-name="two">
                    <div data-toggle-name="nested" data-toggle-accordion-exclude></div>
                    <button data-toggle-target="nested" data-toggle-accordion-exclude></button>
                </div>
                <button data-toggle-target="two"></button>
                <div data-toggle-name="three"></div>
                <button data-toggle-target="three"></button>
            </div>
        `);
            const [, button2, button3] = document.querySelectorAll('button');

            // Act & Assert
            setupToggle();
            TestUtils.click(button3);
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Act & Assert
            TestUtils.click(button2);
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();
        });

    });

    describe('keydown', () => {

        it('should toggle section on pressing \'enter\'', () => {
            // Arrange
            TestUtils.setBodyHtml(`
            <div data-toggle-name="test" class="is-hidden"></div>
            <button data-toggle-target="test"></button>
        `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            const event = new KeyboardEvent('keydown', {
                key: 'Enter'
            });
            button.dispatchEvent(event);

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();
        });

        it('should toggle accordion section on pressing \'enter\'', () => {
            // Arrange
            TestUtils.setBodyHtml(`
                <div data-toggle-accordion>
                    <div data-toggle-name="one" class="is-hidden"></div>
                    <button data-toggle-target="one"></button>
                    <div data-toggle-name="two"></div>
                    <button data-toggle-target="two"></button>
                    <div data-toggle-name="three"></div>
                    <button data-toggle-target="three"></button>
                    <div data-toggle-name="four"></div>
                    <button data-toggle-target="four"></button>
                </div>
            `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            const event = new KeyboardEvent('keydown', {
                key: 'Enter'
            });
            button.dispatchEvent(event);

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();
        });

        it('should toggle accordion section on pressing \'enter\'', () => {
            // Arrange
            TestUtils.setBodyHtml(`
                <div data-toggle-accordion>
                    <div data-toggle-name="one" class="is-hidden">
                        <button data-toggle-target="one"></button>
                        <input />
                    </div>
                    <div data-toggle-name="two">
                        <button data-toggle-target="two"></button>
                    </div>
                </div>
            `);
            const button = document.querySelector('button');

            // Act
            setupToggle();
            const event = new KeyboardEvent('keydown', {
                key: 'Enter'
            });
            button.dispatchEvent(event);

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();
        });

        it('should focus on the next section when pressing \'tab\'', () => {
            // Arrange
            TestUtils.setBodyHtml(`
                <div data-toggle-accordion>
                    <div data-toggle-name="one" class="is-hidden">
                        <button data-toggle-target="one"></button>
                        <input name="one" />
                    </div>
                    <div data-toggle-name="two">
                        <button data-toggle-target="two"></button>
                        <input name="two" />
                    </div>
                </div>
            `);
            const [button1, button2] = document.querySelectorAll('button');

            // Act
            setupToggle();
            const tabEvent = new KeyboardEvent('keydown', {
                key: 'Tab',
                shiftKey: false
            });
            button1.dispatchEvent(tabEvent);

            // Assert
            expect(document.activeElement).toEqual(button2);
        });

        it('should not focus on the next section when pressing \'tab\' and the current section is visible', () => {
            // Arrange
            TestUtils.setBodyHtml(`
                <div data-toggle-accordion data-toggle-first-section-expanded>
                    <div data-toggle-name="one">
                        <button data-toggle-target="one"></button>
                        <input name="one" />
                    </div>
                    <div data-toggle-name="two" class="is-hidden">
                        <button data-toggle-target="two"></button>
                        <input name="two" />
                    </div>
                </div>
            `);
            const [button1, button2] = document.querySelectorAll('button');

            // Act
            setupToggle();
            const tabEvent = new KeyboardEvent('keydown', {
                key: 'Tab',
                shiftKey: false
            });
            button1.dispatchEvent(tabEvent);

            // Assert
            expect(document.activeElement).not.toEqual(button2);
        });

        it('should focus on the previous section when pressing \'shift\' & \'tab\'', () => {
            // Arrange
            TestUtils.setBodyHtml(`
                <div data-toggle-accordion>
                    <div data-toggle-name="one" class="is-hidden">
                        <button data-toggle-target="one"></button>
                        <input name="one" />
                    </div>
                    <div data-toggle-name="two">
                        <button data-toggle-target="two"></button>
                        <input name="two" />
                    </div>
                    <div data-toggle-name="three" class="is-hidden">
                        <button data-toggle-target="three"></button>
                        <input name="three" />
                    </div>
                </div>
            `);
            const [button1, button2] = document.querySelectorAll('button');

            // Act
            setupToggle();
            const tabEvent = new KeyboardEvent('keydown', {
                key: 'Tab',
                shiftKey: true
            });
            button2.dispatchEvent(tabEvent);

            // Assert
            expect(document.activeElement).toEqual(button1);
        });

        it('should not focus on the previous section when pressing \'shift\' & \'tab\' and the previous section is visible', () => {
            // Arrange
            TestUtils.setBodyHtml(`
                <div data-toggle-accordion data-toggle-first-section-expanded>
                    <div data-toggle-name="one">
                        <button data-toggle-target="one"></button>
                        <input name="one" />
                    </div>
                    <div data-toggle-name="two" class="is-hidden">
                        <button data-toggle-target="two"></button>
                        <input name="two" />
                    </div>
                    <div data-toggle-name="three" class="is-hidden">
                        <button data-toggle-target="three"></button>
                        <input name="three" />
                    </div>
                </div>
            `);
            const [button1, button2] = document.querySelectorAll('button');

            // Act
            setupToggle();
            const tabEvent = new KeyboardEvent('keydown', {
                key: 'Tab',
                shiftKey: true
            });
            button2.dispatchEvent(tabEvent);

            // Assert
            // There is an issue here with jsdom. The element of focus will be <input name="one" />,
            // however the test incorrectly returns the full body.
            expect(document.activeElement).not.toEqual(button1);
        });

    });

});
