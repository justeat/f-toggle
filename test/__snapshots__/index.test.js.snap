// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupToggle accordion should display only the first item on initialisation 1`] = `
"
            <div data-toggle-accordion=\\"\\" data-toggle-first-section-expanded=\\"\\">
                <div data-toggle-name=\\"one\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"two\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should have all items collapsed on initialisation 1`] = `
"
            <div data-toggle-accordion=\\"\\">
                <div data-toggle-name=\\"one\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"two\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should hide all other open elements, and toggle the element clicked 1`] = `
"
            <div data-toggle-accordion=\\"\\">
                <div data-toggle-name=\\"one\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"two\\"></button>
                <div data-toggle-name=\\"three\\" class=\\"\\"></div>
                <button data-toggle-target=\\"three\\"></button>
                <div data-toggle-name=\\"four\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"four\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should hide all other open elements, and toggle the element clicked 2`] = `
"
            <div data-toggle-accordion=\\"\\">
                <div data-toggle-name=\\"one\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" class=\\"\\"></div>
                <button data-toggle-target=\\"two\\"></button>
                <div data-toggle-name=\\"three\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"three\\"></button>
                <div data-toggle-name=\\"four\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"four\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should not apply accordion behaviour to elements with 'data-toggle-accordion-exclude' 1`] = `
"
            <div data-toggle-accordion=\\"\\" data-toggle-class=\\"accordion--hide\\">
                <div data-toggle-name=\\"one\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" data-toggle-accordion-exclude=\\"\\"></div>
                <button data-toggle-target=\\"two\\" data-toggle-accordion-exclude=\\"\\"></button>
                <div data-toggle-name=\\"three\\" class=\\"\\"></div>
                <button data-toggle-target=\\"three\\"></button>
                <div data-toggle-name=\\"four\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"four\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should not apply accordion behaviour to elements with 'data-toggle-accordion-exclude' 2`] = `
"
            <div data-toggle-accordion=\\"\\" data-toggle-class=\\"accordion--hide\\">
                <div data-toggle-name=\\"one\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" data-toggle-accordion-exclude=\\"\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"two\\" data-toggle-accordion-exclude=\\"\\"></button>
                <div data-toggle-name=\\"three\\" class=\\"\\"></div>
                <button data-toggle-target=\\"three\\"></button>
                <div data-toggle-name=\\"four\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"four\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should not apply accordion behaviour to elements with 'data-toggle-accordion-exclude' 3`] = `
"
            <div data-toggle-accordion=\\"\\" data-toggle-class=\\"accordion--hide\\">
                <div data-toggle-name=\\"one\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" class=\\"\\">
                    <div data-toggle-name=\\"nested\\" data-toggle-accordion-exclude=\\"\\"></div>
                    <button data-toggle-target=\\"nested\\" data-toggle-accordion-exclude=\\"\\"></button>
                </div>
                <button data-toggle-target=\\"two\\"></button>
                <div data-toggle-name=\\"three\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"three\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should not apply accordion behaviour to elements with 'data-toggle-accordion-exclude' 4`] = `
"
            <div data-toggle-accordion=\\"\\" data-toggle-class=\\"accordion--hide\\">
                <div data-toggle-name=\\"one\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" class=\\"\\">
                    <div data-toggle-name=\\"nested\\" data-toggle-accordion-exclude=\\"\\" class=\\"is-hidden\\"></div>
                    <button data-toggle-target=\\"nested\\" data-toggle-accordion-exclude=\\"\\"></button>
                </div>
                <button data-toggle-target=\\"two\\"></button>
                <div data-toggle-name=\\"three\\" class=\\"accordion--hide\\"></div>
                <button data-toggle-target=\\"three\\"></button>
            </div>
        "
`;

exports[`setupToggle accordion should update the button selected class on selecting new section 1`] = `
"
            <div data-toggle-accordion=\\"\\">
                <div data-toggle-name=\\"one\\" class=\\"is-hidden\\"></div>
                <button data-toggle-target=\\"one\\"></button>
                <div data-toggle-name=\\"two\\" class=\\"\\"></div>
                <button data-toggle-target=\\"two\\"></button>
            </div>
        "
`;

exports[`setupToggle hide method adds custom class 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"toggled\\"></div>
            <button data-toggle-target=\\"hide:test\\" data-toggle-class=\\"toggled\\"></button>
        "
`;

exports[`setupToggle hide method adds hidden class 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"is-hidden\\"></div>
            <button data-toggle-target=\\"hide:test\\"></button>
        "
`;

exports[`setupToggle hide method can handle multiple specified targets 1`] = `
"
            <div data-toggle-name=\\"one\\" class=\\"is-hidden\\"></div>
            <div data-toggle-name=\\"two\\" class=\\"is-hidden\\"></div>
            <button data-toggle-target=\\"hide:one hide:two\\"></button>
        "
`;

exports[`setupToggle hide method can handle multiple targets 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"is-hidden\\"></div>
            <div data-toggle-name=\\"test\\" class=\\"is-hidden\\"></div>
            <button data-toggle-target=\\"hide:test\\"></button>
        "
`;

exports[`setupToggle hide method does nothing when element is already hidden 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"is-hidden\\"></div>
            <button data-toggle-target=\\"hide:test\\"></button>
        "
`;

exports[`setupToggle keydown should toggle accordion section on pressing 'enter' 1`] = `
"
                <div data-toggle-accordion=\\"\\">
                    <div data-toggle-name=\\"one\\" class=\\"\\"></div>
                    <button data-toggle-target=\\"one\\"></button>
                    <div data-toggle-name=\\"two\\" class=\\"is-hidden\\"></div>
                    <button data-toggle-target=\\"two\\"></button>
                    <div data-toggle-name=\\"three\\" class=\\"is-hidden\\"></div>
                    <button data-toggle-target=\\"three\\"></button>
                    <div data-toggle-name=\\"four\\" class=\\"is-hidden\\"></div>
                    <button data-toggle-target=\\"four\\"></button>
                </div>
            "
`;

exports[`setupToggle keydown should toggle accordion section on pressing 'enter' 2`] = `
"
                <div data-toggle-accordion=\\"\\">
                    <div data-toggle-name=\\"one\\" class=\\"\\">
                        <button data-toggle-target=\\"one\\"></button>
                        <input>
                    </div>
                    <div data-toggle-name=\\"two\\" class=\\"is-hidden\\">
                        <button data-toggle-target=\\"two\\"></button>
                    </div>
                </div>
            "
`;

exports[`setupToggle keydown should toggle section on pressing 'enter' 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"\\"></div>
            <button data-toggle-target=\\"test\\"></button>
        "
`;

exports[`setupToggle mixed toggles shows and hides elements correctly 1`] = `
"
            <div data-toggle-name=\\"one\\" class=\\"is-hidden\\"></div>
            <div data-toggle-name=\\"two\\" class=\\"\\"></div>
            <div data-toggle-name=\\"three\\" class=\\"\\"></div>
            <div data-toggle-name=\\"four\\" class=\\"is-hidden\\"></div>
            <button data-toggle-target=\\"one two show:three hide:four\\"></button>
        "
`;

exports[`setupToggle show method can handle multiple specified targets 1`] = `
"
            <div data-toggle-name=\\"one\\" class=\\"\\"></div>
            <div data-toggle-name=\\"two\\" class=\\"\\"></div>
            <button data-toggle-target=\\"show:one show:two\\"></button>
        "
`;

exports[`setupToggle show method can handle multiple targets 1`] = `
"
            <div data-toggle-name=\\"test\\"></div>
            <div data-toggle-name=\\"test\\" class=\\"\\"></div>
            <button data-toggle-target=\\"show:test\\"></button>
        "
`;

exports[`setupToggle show method does nothing when element is already shown 1`] = `
"
            <div data-toggle-name=\\"test\\"></div>
            <button data-toggle-target=\\"show:test\\"></button>
        "
`;

exports[`setupToggle show method removes custom class 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"\\"></div>
            <button data-toggle-target=\\"show:test\\" data-toggle-class=\\"toggled\\"></button>
        "
`;

exports[`setupToggle show method removes hidden class 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"\\"></div>
            <button data-toggle-target=\\"show:test\\"></button>
        "
`;

exports[`setupToggle toggle method adds custom class when specified 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"toggled\\"></div>
            <button data-toggle-target=\\"test\\" data-toggle-class=\\"toggled\\"></button>
        "
`;

exports[`setupToggle toggle method adds hidden class 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"is-hidden\\"></div>
            <button data-toggle-target=\\"test\\"></button>
        "
`;

exports[`setupToggle toggle method can handle multiple specified targets 1`] = `
"
            <div data-toggle-name=\\"one\\" class=\\"\\"></div>
            <div data-toggle-name=\\"two\\" class=\\"is-hidden\\"></div>
            <button data-toggle-target=\\"one two\\"></button>
        "
`;

exports[`setupToggle toggle method can handle multiple targets 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"is-hidden\\"></div>
            <div data-toggle-name=\\"test\\" class=\\"\\"></div>
            <button data-toggle-target=\\"test\\"></button>
        "
`;

exports[`setupToggle toggle method removes custom class when specified 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"\\"></div>
            <button data-toggle-target=\\"test\\" data-toggle-class=\\"toggled\\"></button>
        "
`;

exports[`setupToggle toggle method removes hidden class 1`] = `
"
            <div data-toggle-name=\\"test\\" class=\\"\\"></div>
            <button data-toggle-target=\\"test\\"></button>
        "
`;
